Writing tools rewrite incorrectly removes links in Mail compose

whsieh opened the pull request:
#### becb5858cc51dc73bee1a780568e46b085f41d89
<pre>
Writing tools rewrite incorrectly removes links in Mail compose
<a href="https://bugs.webkit.org/show_bug.cgi?id=281718">https://bugs.webkit.org/show_bug.cgi?id=281718</a>
<a href="https://rdar.apple.com/138041439">rdar://138041439</a>

Reviewed by Richard Robinson.

Using rewrite currently causes any links in Mail compose to be replaced by a span element that has
the same text color, font and underlines as the link, but (importantly) doesn&apos;t contain the actual
link `href`. This happens because the extracted attributed string for WritingTools doesn&apos;t contain
any link URL attributes. To fix this, we teach `WebCore::editingAttributedString` to detect when
text is contained inside of a link, extract the URL of the link, and add it as an attributed string
attribute corresponding to `NSLinkAttributeName`.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(hasAncestorQualifyingForWritingToolsPreservation):

Drive-by fix: use `find()` to avoid a redundant hash lookup.

(enclosingLinkElement):

Add a helper method to return whether a given node is contained within an enclosing link (determined
via `Node::isLink()`). To avoid redundant tree traversal in the case where text is not inside of a
link, we cache the enclosing link result (which may be null) for each node that is both not a link,
and a parent of some text node in the DOM.

(updateAttributes):
(WebCore::editingAttributedString):

See above for more details.

(updateAttributesForStyle): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingWithImage)):

Drive-by fix: typo in a FIXME comment.

(TEST(WritingTools, ProofreadingWithLinks)):

Add a new API test to exercise the change, by verifying that the extracted text context contains
link URLs in the attributed text.

Canonical link: <a href="https://commits.webkit.org/285398@main">https://commits.webkit.org/285398@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/764ab304913b1f07cf5a36b6a69c47e2dec1e106

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72498 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76682 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23707 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23529 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57070 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15587 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75565 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46988 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62423 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37502 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43639 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19890 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22057 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20247 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78394 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16739 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19389 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62435 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16014 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13072 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6711 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47717 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2502 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48786 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50076 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48529 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2613,8 +2640,21 @@ AttributedString editingAttributedString(const SimpleRange& range, OptionSet<Inc
+        if (linkURL.isEmpty())
+            [attributes removeObjectForKey:NSLinkAttributeName];
+        else
+            [attributes setObject:(NSURL *)linkURL forKey:NSLinkAttributeName];
```
maybe this would be better off going in `updateAttributesForStyle`? since that’s where all the other attributes that have the pattern of removeObjectForKey / setObject are in, though doesn’t really matter much

whsieh replied:
We could — but we should then rename `updateAttributesForStyle`, since this link attribute doesn't depend on style at all (just DOM attributes and DOM structure). I'll rename it to `updateAttributes` and pull the logic into there.

Thanks for the review!

whsieh committed:
Writing tools rewrite incorrectly removes links in Mail compose
https://bugs.webkit.org/show_bug.cgi?id=281718
rdar://138041439

Reviewed by Richard Robinson.

Using rewrite currently causes any links in Mail compose to be replaced by a span element that has
the same text color, font and underlines as the link, but (importantly) doesn't contain the actual
link `href`. This happens because the extracted attributed string for WritingTools doesn't contain
any link URL attributes. To fix this, we teach `WebCore::editingAttributedString` to detect when
text is contained inside of a link, extract the URL of the link, and add it as an attributed string
attribute corresponding to `NSLinkAttributeName`.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(hasAncestorQualifyingForWritingToolsPreservation):

Drive-by fix: use `find()` to avoid a redundant hash lookup.

(enclosingLinkElement):

Add a helper method to return whether a given node is contained within an enclosing link (determined
via `Node::isLink()`). To avoid redundant tree traversal in the case where text is not inside of a
link, we cache the enclosing link result (which may be null) for each node that is both not a link,
and a parent of some text node in the DOM.

(updateAttributes):
(WebCore::editingAttributedString):

See above for more details.

(updateAttributesForStyle): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingWithImage)):

Drive-by fix: typo in a FIXME comment.

(TEST(WritingTools, ProofreadingWithLinks)):

Add a new API test to exercise the change, by verifying that the extracted text context contains
link URLs in the attributed text.

Canonical link: https://commits.webkit.org/285398@main
